pkg/ignore/dockerignore.go

```diff
@@ -0,0 +1,242 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ignore
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// DockerIgnoreFile defines the name of the ignore file
+const DockerIgnoreFile = ".dockerignore"
+
+var (
+	_            Ignorer = &defaultIgnorer{}
+	excludedDirs         = sets.NewString()
+	includesDirs         = sets.NewString()
+	directory    string
+)
+
+// dockerIgnorer processes .dockerignore and use the extracted patterns for ignoring files
+type dockerIgnorer struct {
+	patterns *sets.String
+}
+
+// NewOrDefault builds and returns the new or default Ignorer interface
+func NewOrDefault(dir string) (Ignorer, error) {
+	directory = dir
+	//Always include the root directory
+	includesDirs.Insert(dir)
+
+	//Ensure the re-init it every time
+	patterns := sets.NewString(DefaultPatterns...)
+
+	ignoreFile := filepath.Join(dir, DockerIgnoreFile)
+	_, err := os.Stat(ignoreFile)
+
+	if os.IsNotExist(err) {
+		return &defaultIgnorer{
+			&patterns,
+		}, nil
+	}
+
+	fr, err := os.Open(ignoreFile)
+
+	switch {
+	case os.IsNotExist(err):
+		return &defaultIgnorer{
+			&patterns,
+		}, nil
+	case err != nil:
+		return &defaultIgnorer{
+			&patterns,
+		}, err
+	}
+
+	scanner := bufio.NewScanner(fr)
+
+	utf8bom := []byte{0xEF, 0xBB, 0xBF}
+	lineNo := 0
+
+	for scanner.Scan() {
+		scannedBytes := scanner.Bytes()
+
+		if lineNo == 0 {
+			scannedBytes = bytes.TrimPrefix(scannedBytes, utf8bom)
+		}
+
+		pattern := string(scannedBytes)
+		pattern = strings.TrimSpace(pattern)
+		lineNo++
+
+		// Ignore comments
+		if strings.HasPrefix("#", pattern) {
+			continue
+		}
+
+		// Ignore Empty lines
+		if pattern == "" {
+			continue
+		}
+
+		// Normalize - clean and slash
+		invert := pattern[0] == '!'
+		if invert {
+			pattern = strings.TrimSpace(pattern[1:])
+		}
+
+		if len(pattern) > 0 {
+			pattern = filepath.Clean(pattern)
+			pattern = filepath.ToSlash(pattern)
+			if len(pattern) > 1 && pattern[0] == '/' {
+				pattern = pattern[1:]
+			}
+		}
+
+		// add invert to the cleaned pattern
+		if invert {
+			pattern = "!" + pattern
+		}
```

Comment:
seems like this logic is canonicalizing the pattern into a particular shape.  I'd say we should encapsulate that as a function, and for each aspect of the cleanup add a comment illustrating it's purpose (e.g. This strings.TrimSpace is to take `! foo` -> `!foo`).

Reply:
it'll also make unit testing much easier.